### PR TITLE
Refactor miner tests and how bundles are ordered to allow easier extensibility

### PR DESCRIPTION
**Pull request by Ruteri**
## 📝 Summary

Moves the `BuilderPayloadAttributes` structure to common library
Refactors miner test setup
Refactors builder setup and how env header is prepared
New channel unit testing utility
Adds two new kinds of algo tests with proper contract calls in preparation of new kinds of testing we want to do in the near future
Adds simulation backend for use in miner algo unit tests

## 📚 References

Rebased https://github.com/flashbots/builder/pull/27

---

* [x] I have seen and agree to [`CONTRIBUTING.md`](https://github.com/flashbots/builder/blob/main/CONTRIBUTING.md)

### CODE REVIEW

**Comment by Ruteri** [comment]
Cc @lmittmann

File: builder/eth_service.go [hunk 1]
```diff
@@ -43,9 +43,8 @@ func NewEthereumService(eth *eth.Ethereum) *EthereumService {
 	return &EthereumService{eth: eth}
 }
 
-func (s *EthereumService) BuildBlock(attrs *BuilderPayloadAttributes, sealedBlockCallback miner.BlockHookFn) error {
-	// Send a request to generate a full block in the background.
-	// The result can be obtained via the returned channel.
+// TODO: multiservice
```
Comment (dvush) [inline builder/eth_service.go:46]: what is multiservice?

File: miner/algo_common.go [hunk 2]
```diff
@@ -291,25 +291,25 @@ func estimatePayoutTxGas(env *environment, sender, receiver common.Address, prv
...
-			return nil, errors.New("not enough funds available")
-		}
-		rec, err := envDiff.commitPayoutTx(amount, sender, receiver, gas, prv, chData)
-		if err != nil {
-			return nil, fmt.Errorf("failed to commit payment tx: %w", err)
-		} else if rec.Status != types.ReceiptStatusSuccessful {
-			return nil, fmt.Errorf("payment tx failed")
-		}
-		return rec, nil
+func applyTx(envDiff *environmentDiff, sender, receiver common.Address, gas uint64, amountWithFees *big.Int, prv *ecdsa.PrivateKey, chData chainData) (*types.Receipt, error) {
```
Comment (dvush) [inline miner/algo_common.go:294]: nitpick: applyTx is too broad, maybe applyPayoutTx?

**Review by dvush — APPROVED** [review]

File: core/types/transaction.go [hunk 3]
```diff
@@ -459,13 +459,39 @@ func (s TxByNonce) Len() int           { return len(s) }
 func (s TxByNonce) Less(i, j int) bool { return s[i].Nonce() < s[j].Nonce() }
 func (s TxByNonce) Swap(i, j int)      { s[i], s[j] = s[j], s[i] }
 
+type _Order interface {
```
[the remark this answers is not in the document]
Comment (Ruteri) [reply core/types/transaction.go:462]: This particular type is used internally by `TxWithMinerFee` so this is I think the more fitting place

File: builder/eth_service.go (hunk 1 above)
Comment (Ruteri) [reply builder/eth_service.go]: A leftover!

File: miner/algo_common.go (hunk 2 above)
Comment (Ruteri) [reply miner/algo_common.go]: Good point, done

**Pull request merged**